### PR TITLE
Static/Dynamic-Bytes Convenience

### LINKED ISSUE #495: Better `set` for `abi.DynamicArray[abi.Byte]`
## Problem

I'd like to put a byte string as the plain TealType.byte in a DynamicArray with `set` (or some other new method?)

Currently the easiest way I've found is either adding the length manually and calling decode or first creating an abi.String and calling `set` then `byte_array.decode(s.encode())`

### PR DESCRIPTION
**Pull request by ahangsu**
Per #498 issue, we created a pair of helper classes `StaticBytes` and `DynamicBytes` in `abi`, and hopefully the implementation resolves #495 too.

~~Some TODOs down the road: make `set` method easier to use by accepting `bytes`/`bytearray` and `pyteal.bytes`.~~

### CODE REVIEW

File: pyteal/ast/abi/array_dynamic.py [hunk 1]
```diff
@@ -102,3 +97,35 @@ def length(self) -> Expr:
...
+    def set(
+        self,
+        values: Union[
+            bytes,
+            bytearray,
+            Expr,
+            Sequence[Byte],
+            DynamicArray[Byte],
+            ComputedValue[DynamicArray[Byte]],
+        ],
```
Comment (ahangsu) [inline pyteal/ast/abi/array_dynamic.py:117]: I deliberately do not allow input to be `str`, for there might be encoding issue ('utf-8'?) here, and the interface of `set` would be not uniform. If one would like to pass in knowledge contained in `str`, one has to convert to `bytes`.

**Review by jasonpaulos — COMMENTED** [review]
The implementations look good, but could you add unit tests for the new cases in `DynamicBytes.set` and `StaticBytes.set`?

**Comment by ahangsu** [comment]
yes, I need to add `set` testcases for both classes, and I need to update `abi.rst` documentation somewhat.

File: pyteal/ast/abi/array_dynamic.py [hunk 2]
```diff
@@ -102,3 +97,35 @@ def length(self) -> Expr:
 
 
 DynamicArray.__module__ = "pyteal.abi"
+
+
+class DynamicBytes(DynamicArray[Byte]):
+    """The convenience class that represents ABI static byte array."""
```
Comment (barnjamin) [inline pyteal/ast/abi/array_dynamic.py:103]:
```suggestion
    """The convenience class that represents ABI dynamic byte array."""
```

**Comment by barnjamin** [comment]
Looks awesome! Love the helpers for assigning setting!

**Review by jasonpaulos — COMMENTED** [review]
While reviewing I noticed that the existing `_encode_string` function has an issue, so I made a bugfix PR here: https://github.com/algorand/pyteal/pull/508

Ideally that PR would be merged before this one.

File: pyteal/ast/abi/array_dynamic_test.py [hunk 3]
```diff
@@ -203,6 +203,78 @@ def test_DynamicArray_set_computed():
...
+    assert expr.type_of() == pt.TealType.none
+    assert not expr.has_return()
+
+    actual, _ = expr.__teal__(options)
+    actual.addIncoming()
+    actual = actual.NormalizeBlocks(actual)
+
+    expected = pt.TealSimpleBlock(
+        [
+            pt.TealOp(None, pt.Op.byte, "0x" + pt.Bytes(test_case).byte_str),
```
[the remark this answers is not in the document]
Comment (ahangsu) [reply pyteal/ast/abi/array_dynamic_test.py:230]: that would work, I will change accordingly

**Review by jasonpaulos — COMMENTED** [review]
Implementation and tests look good! My only comment is, can you add docstrings to the `set` methods, like this one from `String`? https://github.com/algorand/pyteal/blob/cabe6b87be35f68128e39795464f1446294992b5/pyteal/ast/abi/string.py#L86-L104

Without these, the `DynamicBytes` and `StaticBytes` `set` methods will show the same docstring as their parent classes, which is no longer complete.

**Pull request merged**